src/H5TSrwlock.h

```diff
@@ -313,6 +383,228 @@ H5TS_rwlock_wrunlock(H5TS_rwlock_t *lock)
     return SUCCEED;
 } /* end H5TS_rwlock_wrunlock() */
 
+#elif defined(__MACH__)
+/*-------------------------------------------------------------------------
+ * Function: H5TS_rwlock_rdlock
+ *
+ * Purpose:  Acquire a read lock
+ *
+ * Return:   Non-negative on success / Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static inline herr_t
+H5TS_rwlock_rdlock(H5TS_rwlock_t *lock)
+{
+    /* Check argument */
+    if (H5_UNLIKELY(NULL == lock))
+        return FAIL;
+
+    /* Acquire the lock's mutex */
+    if (H5_UNLIKELY(pthread_mutex_lock(&lock->mutex)))
+        return FAIL;
+
+    /* Check for writers */
+    if (lock->writers || lock->write_waiters) {
+        /* Read waiting */
+        lock->read_waiters++;
+
+        /* Wait for writers */
+        do {
+            if (H5_UNLIKELY(pthread_cond_wait(&lock->read_cv, &lock->mutex))) {
+                pthread_mutex_unlock(&lock->mutex);
+                return FAIL;
+            }
+        } while (lock->writers || lock->write_waiters);
+
+        /* Read not waiting any longer */
+        lock->read_waiters--;
+    }
+
+    /* Increment # of readers */
+    lock->readers++;
+
+    /* Release mutex */
+    if (H5_UNLIKELY(pthread_mutex_unlock(&lock->mutex)))
+        return FAIL;
+
+    return SUCCEED;
+} /* end H5TS_rwlock_rdlock() */
+
+/*-------------------------------------------------------------------------
+ * Function: H5TS_rwlock_rdunlock
+ *
+ * Purpose:  Release a read lock
+ *
+ * Return:   Non-negative on success / Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static inline herr_t
+H5TS_rwlock_rdunlock(H5TS_rwlock_t *lock)
+{
+    /* Check argument */
+    if (H5_UNLIKELY(NULL == lock))
+        return FAIL;
+
+    /* Acquire the lock's mutex */
+    if (H5_UNLIKELY(pthread_mutex_lock(&lock->mutex)))
+        return FAIL;
+
+    /* Decrement # of readers */
+    lock->readers--;
+
+    /* Check for waiting writers when last readers */
+    if (lock->write_waiters && 0 == lock->readers)
+        if (H5_UNLIKELY(pthread_cond_signal(&lock->write_cv))) {
+            pthread_mutex_unlock(&lock->mutex);
+            return FAIL;
+        }
+
+    /* Release mutex */
+    if (H5_UNLIKELY(pthread_mutex_unlock(&lock->mutex)))
+        return FAIL;
+
+    return SUCCEED;
+} /* end H5TS_rwlock_rdunlock() */
+
+/*-------------------------------------------------------------------------
+ * Function: H5TS_rwlock_wrlock
+ *
+ * Purpose:  Acquire a write lock
+ *
+ * Return:   Non-negative on success / Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static inline herr_t
+H5TS_rwlock_wrlock(H5TS_rwlock_t *lock)
```

Comment:
Where does this routine check to see if the thread already owns the lock? Or did I misunderstand the comment about mac pthread?

Reply:
The MacOS version of pthread_rwlock* doesn't obey the POSIX standard's behavior, so these routines re-implement POSIX-compliant R/W locks for it.

POSIX R/W locks are non-recursive, so there's no ownership tracking.

Reply:
Yes, you are correct.   It's desirable because having identical behavior on all platforms is important for developer sanity.  :-)    I had algorithms that "worked" for a long time while I was developing on my Mac laptop, but then started deadlocking when I ran them in the Github CI.   Took a **long, long** time to track that down...

Reply:
So it should work in production with the Mac pthread_rwlock correct? Maybe only use the emulated version in developer mode? Though if there's no performance difference it should be fine

Reply:
Also meant that pthread_rwlock_trywrlock() could succeed on the Mac and acquire the lock, but not acquire the lock on other platforms.   Which was another way to get divergent behavior across platforms...

Reply:
> So it should work in production with the Mac pthread_rwlock correct? Maybe only use the emulated version in developer mode? Though if there's no performance difference it should be fine

Hypothetically, but I believe that it's going to be a while before we trust the code _that_ much.

Reply:
The pthread_rwlock_trywrlock() behavior is really the killer reason for keeping the code the same in production as well as developer mode.